pennylane/devices/device_api.py

```diff
@@ -463,7 +489,68 @@ def execute_fn(tapes):
                 return self.preprocess(execution_config)[0]
             return self.preprocess()[0]
 
-        return TransformProgram()
+        if not self.capabilities:
+            # The capabilities are required to construct a default transform program.
+            return TransformProgram()
+
+        if not execution_config:
+            execution_config = ExecutionConfig()
+
+        program = TransformProgram()
+
+        # First handle mid-circuit measurements because it may add wires. At this point we
+        # should assume that the mcm method is already validated and resolved.
+        if execution_config.mcm_config.mcm_method == "deferred":
+            program.add_transform(
+                qml.transforms.defer_measurements,
+                allow_postselect=self.capabilities.supports_operation("Projector"),
+            )
+        elif execution_config.mcm_config.mcm_method == "one-shot":
+            program.add_transform(
+                qml.transforms.dynamic_one_shot,
+                postselect_mode=execution_config.mcm_config.postselect_mode,
+            )
+
+        capabilities_analytic = self.capabilities.filter(finite_shots=False)
+        capabilities_shots = self.capabilities.filter(finite_shots=True)
+
+        program.add_transform(
+            decompose,
+            stopping_condition=lambda o: capabilities_analytic.supports_operation(o.name),
+            stopping_condition_shots=lambda o: capabilities_shots.supports_operation(o.name),
```

Comment:
Would it make sense to move the `o.name` part into `supports_operation`?

Reply:
In retrospect that's probably what I should've done but to make such a change now means another round of updates in PL/Catalyst/Lightning again. 🤦🏻

Reply:
Could we have `supports_operation(op : str| type[Operator])` where we allow both? Shouldn't necessarily break things on the consumer side then.

Reply:
Done